File: frontend/src/lib/query-spec.ts
/**
 * Каноническое описание выборки — зеркало серверного `QuerySpec`.
 *
 * ТЗ требует, чтобы «Списком» и «На карте» показывали одну и ту же выборку.
 * Значит, состояние фильтров не может жить внутри компонента: при
 * переключении представления оно бы сбрасывалось или расходилось. Оно живёт в
 * адресной строке, и этот модуль — единственное место, где описано, как оно
 * туда кладётся и как оттуда читается.
 *
 * Побочная выгода: ссылку можно переслать коллеге, и он увидит ровно ту же
 * выборку, а кнопка «назад» в браузере восстанавливает предыдущую.
 */

import { RISK_LEVELS, type RiskLevel } from "@/lib/risk";

export const SORT_FIELDS = ["risk", "amount", "relevance", "name"] as const;
export type SortField = (typeof SORT_FIELDS)[number];

export const SORT_ORDERS = ["asc", "desc"] as const;
export type SortOrder = (typeof SORT_ORDERS)[number];

export const OBJECT_TYPES = [
  "territory",
  "contract",
  "subsidy_recipient",
  "ppp_project",
  "expertise_object",
  "organization",
] as const;
export type ObjectType = (typeof OBJECT_TYPES)[number];

export const SORT_FIELD_LABELS: Record<SortField, string> = {
  risk: "По риску",
  amount: "По сумме",
  relevance: "По актуальности",
  name: "По названию",
};

export const OBJECT_TYPE_LABELS: Record<ObjectType, string> = {
  territory: "Территории",
  contract: "Договоры",
  subsidy_recipient: "Получатели субсидий",
  ppp_project: "Проекты ГЧП",
  expertise_object: "Объекты экспертизы",
  organization: "Организации",
};

export interface QuerySpec {
  dateFrom: string | null;
  dateTo: string | null;
  year: number | null;

  territoryCodes: string[];
  includeChildTerritories: boolean;

  objectTypes: ObjectType[];
  layers: string[];
  industries: string[];
  statuses: string[];

  amountMin: number | null;
  amountMax: number | null;

  riskLevels: RiskLevel[];
  completenessMin: number | null;
  completenessMax: number | null;
  onlyCategoryA: boolean;

  search: string | null;

  sort: SortField;
  order: SortOrder;
  page: number;
  pageSize: number;
}

/**
 * Значения по умолчанию.
 *
 * Пустая выборка означает «всё, что есть», а не «ничего»: пользователь,
 * впервые открывший экран, должен увидеть данные, а не просьбу настроить
 * фильтры.
 *
 * Уровень «нет данных» включён намеренно. Убрать его из умолчаний — значит
 * молча спрятать неизмеренные объекты от того, кто фильтры не трогал.
 */
export const DEFAULT_QUERY_SPEC: QuerySpec = {
  dateFrom: null,
  dateTo: null,
  year: null,

  territoryCodes: [],
  includeChildTerritories: true,

  objectTypes: [],
  layers: [],
  industries: [],
  statuses: [],

  amountMin: null,
  amountMax: null,

  riskLevels: [...RISK_LEVELS],
  completenessMin: null,
  completenessMax: null,
  onlyCategoryA: false,

  search: null,

  sort: "risk",
  order: "desc",
  page: 1,
  pageSize: 25,
};

/** Имена в адресной строке — snake_case, как на сервере. */
const PARAM_NAMES: Record<keyof QuerySpec, string> = {
  dateFrom: "date_from",
  dateTo: "date_to",
  year: "year",
  territoryCodes: "territory_codes",
  includeChildTerritories: "include_child_territories",
  objectTypes: "object_types",
  layers: "layers",
  industries: "industries",
  statuses: "statuses",
  amountMin: "amount_min",
  amountMax: "amount_max",
  riskLevels: "risk_levels",
  completenessMin: "completeness_min",
  completenessMax: "completeness_max",
  onlyCategoryA: "only_category_a",
  search: "search",
  sort: "sort",
  order: "order",
  page: "page",
  pageSize: "page_size",
};

function sameList(a: readonly unknown[], b: readonly unknown[]): boolean {
  if (a.length !== b.length) return false;
  const left = [...a].map(String).sort();
  const right = [...b].map(String).sort();
  return left.every((value, index) => value === right[index]);
}

/**
 * Превратить выборку в параметры адресной строки.
 *
 * Значения, совпадающие с умолчанием, опускаются: ссылка должна оставаться
 * читаемой, а не тащить два десятка параметров, ничего не меняющих.
 */
export function toSearchParams(spec: QuerySpec): URLSearchParams {
  const params = new URLSearchParams();

  for (const key of Object.keys(PARAM_NAMES) as (keyof QuerySpec)[]) {
    const value = spec[key];
    const fallback = DEFAULT_QUERY_SPEC[key];
    const name = PARAM_NAMES[key];

    if (Array.isArray(value)) {
      if (sameList(value, fallback as readonly unknown[])) continue;
      params.set(name, value.join(","));
      continue;
    }

    if (value === fallback || value === null) continue;
    params.set(name, String(value));
  }

  return params;
}

function parseNumber(raw: string | null): number | null {
  if (raw === null || raw.trim() === "") return null;
  const parsed = Number(raw);
  return Number.isFinite(parsed) ? parsed : null;
}

function parseList<T extends string>(raw: string | null, allowed: readonly T[]): T[] | null {
  if (raw === null) return null;
  if (raw === "") return [];
  const allowedSet = new Set<string>(allowed);
  return raw
    .split(",")
    .map((part) => part.trim())
    .filter((part): part is T => allowedSet.has(part));
}

/**
 * Прочитать выборку из адресной строки.
 *
 * Неизвестные и повреждённые значения игнорируются и заменяются умолчанием:
 * ссылка, которую кто-то поправил руками или обвешал метками рекламной
 * аналитики, не должна ронять страницу.
 */
export function fromSearchParams(params: URLSearchParams | null | undefined): QuerySpec {
  if (!params) return { ...DEFAULT_QUERY_SPEC };

  const get = (key: keyof QuerySpec) => params.get(PARAM_NAMES[key]);

  const risk = parseList(get("riskLevels"), RISK_LEVELS);
  const sortRaw = get("sort");
  const orderRaw = get("order");

  const page = parseNumber(get("page"));
  const pageSize = parseNumber(get("pageSize"));

  return {
    dateFrom: get("dateFrom"),
    dateTo: get("dateTo"),
    year: parseNumber(get("year")),

    territoryCodes: params.has(PARAM_NAMES.territoryCodes)
      ? (get("territoryCodes") ?? "").split(",").filter(Boolean)
      : [...DEFAULT_QUERY_SPEC.territoryCodes],
    includeChildTerritories: get("includeChildTerritories") !== "false",

    objectTypes: parseList(get("objectTypes"), OBJECT_TYPES) ?? [
      ...DEFAULT_QUERY_SPEC.objectTypes,
    ],
    layers: params.has(PARAM_NAMES.layers)
      ? (get("layers") ?? "").split(",").filter(Boolean)
      : [...DEFAULT_QUERY_SPEC.layers],
    industries: params.has(PARAM_NAMES.industries)
      ? (get("industries") ?? "").split(",").filter(Boolean)
      : [...DEFAULT_QUERY_SPEC.industries],
    statuses: params.has(PARAM_NAMES.statuses)
      ? (get("statuses") ?? "").split(",").filter(Boolean)
      : [...DEFAULT_QUERY_SPEC.statuses],

    amountMin: parseNumber(get("amountMin")),
    amountMax: parseNumber(get("amountMax")),

    /*
      Пустой список уровней означал бы заведомо пустую выборку, и пользователь
      не понял бы, почему ничего не видно. Такое состояние трактуется как
      «фильтр не задан».
    */
    riskLevels: risk && risk.length > 0 ? risk : [...DEFAULT_QUERY_SPEC.riskLevels],
    completenessMin: parseNumber(get("completenessMin")),
    completenessMax: parseNumber(get("completenessMax")),
    onlyCategoryA: get("onlyCategoryA") === "true",

    search: get("search")?.trim() || null,

    sort: SORT_FIELDS.includes(sortRaw as SortField)
      ? (sortRaw as SortField)
      : DEFAULT_QUERY_SPEC.sort,
    order: SORT_ORDERS.includes(orderRaw as SortOrder)
      ? (orderRaw as SortOrder)
      : DEFAULT_QUERY_SPEC.order,
    page: page && page >= 1 ? Math.floor(page) : 1,
    pageSize: pageSize && pageSize >= 1 ? Math.min(Math.floor(pageSize), 200) : 25,
  };
}

/** Отличается ли выборка от умолчания. Страница и сортировка фильтром не считаются. */
export function hasActiveFilters(spec: QuerySpec): boolean {
  const params = toSearchParams(spec);
  for (const ignored of ["page", "page_size", "sort", "order"]) {
    params.delete(ignored);
  }
  return [...params.keys()].length > 0;
}

export interface FilterChip {
  key: keyof QuerySpec;
  label: string;
  value: string;
}

/** Чипы активных фильтров над списком и картой. */
export function activeFilterChips(spec: QuerySpec): FilterChip[] {
  const chips: FilterChip[] = [];

  if (spec.year !== null) {
    chips.push({ key: "year", label: "Период", value: String(spec.year) });
  } else if (spec.dateFrom || spec.dateTo) {
    chips.push({
      key: "dateFrom",
      label: "Период",
      value: `${spec.dateFrom ?? "…"} — ${spec.dateTo ?? "…"}`,
    });
  }

  if (spec.territoryCodes.length > 0) {
    chips.push({
      key: "territoryCodes",
      label: "Территория",
      value: `выбрано: ${spec.territoryCodes.length}`,
    });
  }

  if (spec.objectTypes.length > 0) {
    chips.push({
      key: "objectTypes",
      label: "Тип объекта",
      value: spec.objectTypes.map((t) => OBJECT_TYPE_LABELS[t]).join(", "),
    });
  }

  if (spec.amountMin !== null || spec.amountMax !== null) {
    const low = spec.amountMin !== null ? spec.amountMin.toLocaleString("ru-RU") : "0";
    const high = spec.amountMax !== null ? spec.amountMax.toLocaleString("ru-RU") : "∞";
    chips.push({ key: "amountMin", label: "Сумма", value: `${low} — ${high} ₸` });
  }

  if (!sameList(spec.riskLevels, DEFAULT_QUERY_SPEC.riskLevels)) {
    chips.push({
      key: "riskLevels",
      label: "Уровень риска",
      value: `выбрано: ${spec.riskLevels.length}`,
    });
  }

  if (spec.completenessMin !== null || spec.completenessMax !== null) {
    const low = Math.round((spec.completenessMin ?? 0) * 100);
    const high = Math.round((spec.completenessMax ?? 1) * 100);
    chips.push({ key: "completenessMin", label: "Полнота данных", value: `${low}% — ${high}%` });
  }

  if (spec.onlyCategoryA) {
    chips.push({ key: "onlyCategoryA", label: "Категория", value: "только категория A" });
  }

  if (spec.search) {
    chips.push({ key: "search", label: "Поиск", value: spec.search });
  }

  return chips;
}

/** Снять один фильтр, вернув его к умолчанию. */
export function clearFilter(spec: QuerySpec, key: keyof QuerySpec): QuerySpec {
  const reset: Partial<QuerySpec> = { [key]: DEFAULT_QUERY_SPEC[key] } as Partial<QuerySpec>;

  // Период задаётся парой полей: снимать его по одному бессмысленно.
  if (key === "dateFrom" || key === "dateTo") {
    reset.dateFrom = null;
    reset.dateTo = null;
  }
  if (key === "amountMin" || key === "amountMax") {
    reset.amountMin = null;
    reset.amountMax = null;
  }
  if (key === "completenessMin" || key === "completenessMax") {
    reset.completenessMin = null;
    reset.completenessMax = null;
  }

  // Снятие любого фильтра возвращает на первую страницу: иначе пользователь
  // окажется на седьмой странице выборки, в которой теперь три результата.
  return { ...spec, ...reset, page: 1 };
}

/** Изменить фильтры, сбросив страницу. */
export function withFilters(spec: QuerySpec, patch: Partial<QuerySpec>): QuerySpec {
  const changesPagination = "page" in patch || "pageSize" in patch;
  return { ...spec, ...patch, page: changesPagination ? (patch.page ?? spec.page) : 1 };
}

File: backend/app/api/queryspec.py
"""Каноническое описание выборки.

ТЗ требует, чтобы «Списком» и «На карте» были двумя представлениями **одной**
выборки, а не двумя страницами с расходящимся состоянием. Единственный надёжный
способ это обеспечить — одно описание фильтров, из которого одинаково строятся
и запрос карты, и запрос списка, и ссылка в адресной строке.

Отсюда три свойства этого модуля:

* фильтры описаны один раз и живут в одном объекте;
* объект умеет превращаться в параметры URL и обратно без потерь, иначе
  кнопка «назад» в браузере восстановит не то состояние;
* сортировка и страница — часть выборки, а не отдельное состояние компонента.

Отдельно про уровень риска. `unknown` — полноправное значение фильтра, и по
умолчанию оно **включено**. Если убрать его из умолчаний, пользователь, не
трогавший фильтры, увидит только измеренные объекты и решит, что видит всё.
"""

from __future__ import annotations

from datetime import date
from enum import StrEnum
from typing import Any, Self
from urllib.parse import urlencode

from pydantic import BaseModel, Field, field_validator, model_validator

from app.risk.core import RiskLevel


class SortField(StrEnum):
    """Поля сортировки из ТЗ, раздел про режим списка."""

    RISK = "risk"
    AMOUNT = "amount"
    RELEVANCE = "relevance"
    """По дате актуальности данных."""

    NAME = "name"


class SortOrder(StrEnum):
    ASC = "asc"
    DESC = "desc"


class ObjectType(StrEnum):
    """Типы объектов, участвующие в выборке.

    Совпадают со слоями данных. Организации входят в список, хотя на карту не
    выводятся: у них нет географии, но искать и открывать их карточку нужно.
    """

    TERRITORY = "territory"
    CONTRACT = "contract"
    SUBSIDY_RECIPIENT = "subsidy_recipient"
    PPP_PROJECT = "ppp_project"
    EXPERTISE_OBJECT = "expertise_object"
    ORGANIZATION = "organization"


class QuerySpec(BaseModel):
    """Выборка: что показать, как отсортировать, какую страницу.

    Значения по умолчанию подобраны так, чтобы пустая выборка означала «всё,
    что есть», а не «ничего»: пользователь, впервые открывший экран, должен
    увидеть данные, а не пустой список с просьбой настроить фильтры.
    """

    model_config = {"extra": "forbid"}

    # --- Период --------------------------------------------------------------

    date_from: date | None = None
    date_to: date | None = None
    year: int | None = Field(default=None, ge=2000, le=2100)

    compare_to_from: date | None = None
    compare_to_to: date | None = None
    """Второй период для сравнения. ТЗ требует кнопку «Сравнить периоды»."""

    # --- Территория ----------------------------------------------------------

    territory_codes: list[str] = Field(default_factory=list)
    include_child_territories: bool = True
    """Выбор области подразумевает её районы — иначе выбор области даёт пусто."""

    # --- Предметные фильтры --------------------------------------------------

    object_types: list[ObjectType] = Field(default_factory=list)
    layers: list[str] = Field(default_factory=list)
    industries: list[str] = Field(default_factory=list)
    sources: list[str] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)

    customer_ids: list[str] = Field(default_factory=list)
    supplier_ids: list[str] = Field(default_factory=list)
    recipient_ids: list[str] = Field(default_factory=list)

    amount_min: float | None = Field(default=None, ge=0)
    amount_max: float | None = Field(default=None, ge=0)

    # --- Риск ----------------------------------------------------------------

    risk_levels: list[RiskLevel] = Field(
        default_factory=lambda: list(RiskLevel),
        description=(
            "По умолчанию включены ВСЕ уровни, включая «нет данных». Иначе "
            "пользователь, не трогавший фильтры, увидит только измеренные "
            "объекты и решит, что видит всё."
        ),
    )
    completeness_min: float | None = Field(default=None, ge=0, le=1)
    completeness_max: float | None = Field(default=None, ge=0, le=1)
    only_category_a: bool = False
    """Только объекты со сработавшим жёстким переопределением уровня."""

    # --- Поиск ---------------------------------------------------------------

    search: str | None = Field(default=None, max_length=255)

    # --- Представление -------------------------------------------------------

    sort: SortField = SortField.RISK
    order: SortOrder = SortOrder.DESC
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=25, ge=1, le=200)

    @field_validator("search")
    @classmethod
    def _strip_search(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None

    @field_validator("territory_codes", "industries", "sources", "statuses", mode="before")
    @classmethod
    def _split_csv(cls, value: object) -> object:
        """Принять и повторяющийся параметр, и список через запятую.

        Браузеры и клиенты передают множественные значения по-разному
        (`?t=a&t=b` против `?t=a,b`). Поддерживаем оба, иначе ссылка,
        собранная вручную, молча теряет часть фильтров.
        """
        if isinstance(value, str):
            return [part.strip() for part in value.split(",") if part.strip()]
        return value

    @model_validator(mode="after")
    def _check_ranges(self) -> Self:
        if self.date_from and self.date_to and self.date_from > self.date_to:
            raise ValueError("начало периода позже его конца")
        if (
            self.amount_min is not None
            and self.amount_max is not None
            and self.amount_min > self.amount_max
        ):
            raise ValueError("нижняя граница суммы больше верхней")
        if (
            self.completeness_min is not None
            and self.completeness_max is not None
            and self.completeness_min > self.completeness_max
        ):
            raise ValueError("нижняя граница полноты больше верхней")
        if not self.risk_levels:
            raise ValueError(
                "не выбран ни один уровень риска — такая выборка всегда пуста; "
                "чтобы снять фильтр, оставьте все уровни включёнными"
            )
        return self

    # --- Представление в URL -------------------------------------------------

    def to_query_params(self) -> dict[str, str]:
        """Параметры адресной строки.

        Значения, равные умолчанию, опускаются: ссылка должна оставаться
        читаемой, а не тащить три десятка параметров, ничего не меняющих.
        """
        defaults = QuerySpec()
        params: dict[str, str] = {}

        for name, value in self.model_dump(exclude_none=True).items():
            default = getattr(defaults, name)
            if isinstance(default, list):
                default_set = {str(item) for item in default}
                current_set = {str(item) for item in value}
                if current_set == default_set:
                    continue
                if value:
                    params[name] = ",".join(str(item) for item in value)
                else:
                    # Пустой список отличается от умолчания и должен
                    # сохраняться в ссылке, иначе фильтр «ничего не выбрано»
                    # восстановится как «выбрано всё».
                    params[name] = ""
                continue

            if value == default:
                continue
            params[name] = str(value)

        return params

    def to_query_string(self) -> str:
        return urlencode(self.to_query_params())

    @classmethod
    def from_query_params(cls, params: dict[str, Any]) -> QuerySpec:
        """Разобрать параметры адресной строки.

        Пустая строка у списочного параметра означает «выбрано пусто», а
        отсутствие параметра — «значение по умолчанию». Разница существенна:
        первое даёт пустую выборку, второе — полную.
        """
        prepared: dict[str, Any] = {}
        list_fields = {
            name
            for name, field in cls.model_fields.items()
            if str(field.annotation).startswith("list")
        }

        for key, raw in params.items():
            if key not in cls.model_fields:
                continue
            if key in list_fields:
                if raw == "" or raw is None:
                    prepared[key] = []
                elif isinstance(raw, str):
                    prepared[key] = [p.strip() for p in raw.split(",") if p.strip()]
                else:
                    prepared[key] = raw
            else:
                prepared[key] = raw

        return cls.model_validate(prepared)

    # --- Удобства ------------------------------------------------------------

    @property
    def offset(self) -> int:
        return (self.page - 1) * self.page_size

    @property
    def includes_unknown_risk(self) -> bool:
        return RiskLevel.UNKNOWN in self.risk_levels

    @property
    def has_active_filters(self) -> bool:
        """Отличается ли выборка от умолчания.

        Нужно интерфейсу: если фильтры не тронуты, показывать чипы и кнопку
        сброса незачем.
        """
        return bool(self.to_query_params().keys() - {"page", "page_size", "sort", "order"})

    def active_filter_chips(self) -> list[tuple[str, str]]:
        """Пары «поле → человекочитаемое значение» для чипов над списком."""
        chips: list[tuple[str, str]] = []

        if self.year:
            chips.append(("Период", str(self.year)))
        elif self.date_from or self.date_to:
            начало = self.date_from.strftime("%d.%m.%Y") if self.date_from else "…"
            конец = self.date_to.strftime("%d.%m.%Y") if self.date_to else "…"
            chips.append(("Период", f"{начало} — {конец}"))

        if self.territory_codes:
            chips.append(("Территория", f"выбрано: {len(self.territory_codes)}"))
        if self.object_types:
            chips.append(("Тип объекта", f"выбрано: {len(self.object_types)}"))
        if self.industries:
            chips.append(("Отрасль", f"выбрано: {len(self.industries)}"))
        if self.statuses:
            chips.append(("Статус", f"выбрано: {len(self.statuses)}"))

        if self.amount_min is not None or self.amount_max is not None:
            низ = f"{self.amount_min:,.0f}" if self.amount_min is not None else "0"
            верх = f"{self.amount_max:,.0f}" if self.amount_max is not None else "∞"
            chips.append(("Сумма", f"{низ} — {верх} ₸"))

        if set(self.risk_levels) != set(RiskLevel):
            names = ", ".join(level.label_ru for level in self.risk_levels)
            chips.append(("Уровень риска", names))

        if self.completeness_min is not None or self.completeness_max is not None:
            низ = f"{(self.completeness_min or 0) * 100:.0f}%"
            верх = f"{(self.completeness_max or 1) * 100:.0f}%"
            chips.append(("Полнота данных", f"{низ} — {верх}"))

        if self.only_category_a:
            chips.append(("Категория", "только категория A"))

        if self.search:
            chips.append(("Поиск", self.search))

        return chips

    def for_page(self, page: int) -> QuerySpec:
        return self.model_copy(update={"page": page})

    def without_pagination(self) -> QuerySpec:
        """Та же выборка без страницы — для карты и агрегатов.

        Карта показывает выборку целиком: пагинация относится только к списку.
        """
        return self.model_copy(update={"page": 1, "page_size": 200})


class PageInfo(BaseModel):
    """Сведения о странице результата."""

    page: int
    page_size: int
    total: int
    total_pages: int

    @classmethod
    def build(cls, spec: QuerySpec, total: int) -> PageInfo:
        pages = (total + spec.page_size - 1) // spec.page_size if total else 0
        return cls(page=spec.page, page_size=spec.page_size, total=total, total_pages=pages)

    @property
    def has_next(self) -> bool:
        return self.page < self.total_pages

    @property
    def has_previous(self) -> bool:
        return self.page > 1
